merkletree/pad.go

```diff
@@ -41,29 +45,41 @@ func NewPAD(policies Policies, key crypto.SigningKey, len int64) (*PAD, error) {
 }
 
 // if policies is nil, the previous policies will be used
-func (pad *PAD) generateNextSTR(policies Policies, m *MerkleTree, epoch uint64) {
+func (pad *PAD) signTreeRoot(m *MerkleTree, epoch uint64) {
 	var prevStrHash []byte
-	if pad.currentSTR == nil {
+	if pad.latestSTR == nil {
 		prevStrHash = make([]byte, crypto.HashSizeByte)
 		if _, err := rand.Read(prevStrHash); err != nil {
 			// panic here since if there is an error, it will break the PAD.
 			panic(err)
 		}
 	} else {
-		prevStrHash = crypto.Digest(pad.currentSTR.sig)
-		if policies == nil {
-			policies = pad.currentSTR.policies
-		}
+		prevStrHash = crypto.Digest(pad.latestSTR.sig)
 	}
-	pad.currentSTR = NewSTR(pad.key, policies, m, epoch, prevStrHash)
+	pad.latestSTR = NewSTR(pad.key, pad.policies, m, epoch, prevStrHash)
 }
 
 func (pad *PAD) updateInternal(policies Policies, epoch uint64) {
 	pad.tree.recomputeHash()
 	m := pad.tree.Clone()
-	pad.generateNextSTR(policies, m, epoch)
-	pad.snapshots[epoch] = pad.currentSTR
+	// create STR with the policies that were actually used in the prev.
+	// Set() operation
+	pad.signTreeRoot(m, epoch)
+	pad.snapshots[epoch] = pad.latestSTR
 	pad.loadedEpochs = append(pad.loadedEpochs, epoch)
+
+	if policies != nil { // update the policies if necessary
+		// TODO use constant time compare from "crypto/subtle" instead (?)
+		vrfKeyChanged := !bytes.Equal(pad.policies.vrfPrivate()[:],
+			policies.vrfPrivate()[:])
```

Comment:
Unless @masomel and @arlolra raise any problems about it, I think it's fine. But if we use constant time compare, it also give a sense of security and it doesn't hurt anybody.